knit/dask_yarn.py

```diff
@@ -96,35 +96,34 @@ def start(self, n_workers=1, cpus=1, memory=2048, checks=True,
         -------
         YARN application ID.
         """
-        c = CondaCreator(channels=self.channels, **(self.conda_pars or {}))
         if self.env is None:
+            c = CondaCreator(channels=self.channels, **(self.conda_pars or {}))
             env_name = 'dask-' + sha1(
-                '-'.join(self.packages).encode()).hexdigest()
+                '-'.join(self.packages + self.channels).encode()).hexdigest()
             env_path = os.path.join(c.conda_envs, env_name)
             if os.path.exists(env_path + '.zip'):
                 # zipfile exists, ready to upload
                 self.env = env_path + '.zip'
             elif os.path.exists(env_path):
                 # environment exists, can zip and upload
-                c.zip_env(env_path)
-                self.env = env_path + '.zip'
+                self.env = zip_path(env_path)
             else:
                 # create env from scratch
                 self.env = c.create_env(env_name=env_name,
                                         packages=self.packages)
         elif not self.env.endswith('.zip'):
             # given env directory, so zip it
-            c.zip_env(self.env)
-            self.env = self.env + '.zip'
+            self.env = zip_path(self.env)
 
         # TODO: memory should not be total available?
-        command = '$PYTHON_BIN $CONDA_PREFIX/bin/dask-worker --nprocs=1 ' \
-                  '--nthreads=%d --memory-limit=%d %s > ' \
-                  '/tmp/worker-log.out 2> /tmp/worker-log.err' % (
-                      cpus, memory * 1e6,
-                      self.local_cluster.scheduler.address)
-
-        app_id = self.knit.start(command, env=self.env,
+        bn = os.path.basename(self.env)
+        pref = bn + '/' + bn[:-4]  # like myenv.zip/myenv
```

Comment:
Why this? It seems to me the environment zip file shouldn't unzip to a directory with `.zip` in the name. I'd expect `unzip myenv.zip` to result in a folder `myenv`, meaning `pref` should be `myenv/bin/python`. This seems to imply that `unzip myenv.zip` results in a folder `myenv.zip` with a nested folder named `myenv`.

Reply:
Either way you probably should be using `os.path.join` and `os.path.splitext` instead of `+` and `[:-4]`.

Reply:
The zip files are always expanded inside a directory, so if you pass in `my.zip` made from directory `my/`, you will get `my.zip/my/files`. The top level could be named anything, but it seems pretty clear to me.
I wouldn't use `path.join`, since the containers will always run on a unix system, whereas it should be possible to launch applications from windows.

Reply:
You can't expand a zip file in the worker root folder, or expand all of them in a folder and then `cd` into that folder? The behavior here is very unexpected (at least to me).

> I wouldn't use path.join, since the containers will always run on a unix system, whereas it should be possible to launch applications from windows.

I'd argue that you should always write what's intended, even if the system it's running on will never need to be cross platform. `os.path.join` means "join some paths together", while `x + '/' + x2...` means concatenate some strings. Only in context can I see that the string concatenation is joining paths.

Reply:
> You can't expand a zip file in the worker root folder, or expand all of them in a folder and then cd into that folder?

No, all you do is give a name to YARN, which takes care of doing the unzipping.

> if the system it's running on will never need to be cross platform.

I mean that we are testing linux/unix and everything is fine, but when this will be used on windows/unix, then path.join will break. In fact, yarn - the server - could be run on windows, I suppose, but probably no one does that.

Reply:
Hmmm, ok. I found very little documentation about this. There's a stack-overflow post saying that [`spark-submit` does the same thing](https://stackoverflow.com/questions/41498365/upload-zip-file-using-archives-option-of-spark-submit-on-yarn), and there's the [yarn docs](https://hadoop.apache.org/docs/r2.7.2/hadoop-yarn/hadoop-yarn-site/WritingYarnApplications.html) which indicate how the local cwd is setup for files but not for zip files (first answer in `FAQs`, can't link to it directly). I still find this behavior surprising and non-intuitive.

> I mean that we are testing linux/unix and everything is fine, but when this will be used on windows/unix, then path.join will break.

Since this is meant to be used on the edge node, I'd expect the resourcemanager and edge node to share the same OS. I've run into enough buggy code doing path processing using string operations instead of path operations that maybe I'm overly wary of this.